### PR TITLE
[object] Fix reproducer crash when input file has no contents

### LINKED ISSUE #818: Reproducer crashes when the link contains a script that is found in the sysroot
The reproducer crashes when the link contains a script that is found in the sysroot.

Reproducer:

```bash
#!/usr/bin/env bash

mkdir -p A/lib64
mkdir -p A/script

cat >1.c <<\EOF
int foo() { return 1; }
EOF

cat >script.t <<\EOF
GROUP(/lib64/lib1.so)
EOF

cat >main.c <<\EOF
int main() { return 0; }
EOF

clang -o lib1.so -target x86_64-unknown-elf 1.c -shared -fPIC
clang -o main.o -target x86_64-unknown-elf main.c -c 
mv lib1.so A/lib64/lib1.so
mv script.t A/script.t

ld.eld --sysroot=A -T A/script.t main.o --reproduce a.tar
rm -rf unpacked && mkdir unpacked
tar -xvf a.tar -C unpacked --strip-components=1
cd unpacked
bash -x response.txt
```

### PR DESCRIPTION
**Pull request by deepakshirkem**
###  Problem
ELD crashes with a segfault when a reproduce tarball contains a linker script that references a file not found in the sysroot.

### Testing
Added test `SysrootScriptCrash` that verifies ELD errors out  gracefully instead of segfaulting when a linker script references an unreadable file.

### Screenshot
Screenshot from 2026-05-12 18-52-19

Fixes #818

cc @quic-seaswara , @parth-07 , @quic-areg

### CODE REVIEW

File: lib/Object/ObjectLinker.cpp [hunk 1]
```diff
@@ -3614,8 +3614,10 @@ void ObjectLinker::addInputFileToTar(InputFile *Ipt, MappingFile::Kind K) {
 bool ObjectLinker::readAndProcessInput(Input *Input, bool IsPostLto) {
   if (Input->isInternal())
     return true;
-  if (!Input->getSize())
+  if (!Input->getSize()) {
     ThisConfig.raise(Diag::input_file_has_zero_size) << Input->decoratedPath();
```
Comment (parth-07) [inline lib/Object/ObjectLinker.cpp:3608]: Are we emitting this diagnostic for non-existing file? If so, that is incorrect. Also with this modification, the link will fail for existing but 0-sized files. Is this an intended behavior change?

Comment (deepakshirkem) [reply lib/Object/ObjectLinker.cpp:3608]: Hi @parth-07, I think you are right in both cases. I will update the checks.

File: test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#---SysrootScriptCrash.test----------- Executable -----------------#
+#BEGIN_COMMENT
+# Verify that ELD does not crash when a linker script references a
+# file that cannot be read (e.g. not found in sysroot).
+# ELD should error out gracefully instead of segfaulting.
+# Fix for issue #818.
```
Comment (quic-areg) [inline test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: github specific references can probably be left out.

Comment (deepakshirkem) [reply test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: Hi @quic-areg,  Thank You ::)). Will remove it.

**Comment by deepakshirkem** [comment]
Hi @parth-07,  I have updated the fix. Here is the verified behavior 
Screenshot from 2026-05-12 21-30-50

Hi @quic-areg , Updated the test. Please go one more time.

File: lib/Object/ObjectLinker.cpp [hunk 3]
```diff
@@ -3614,8 +3614,13 @@ void ObjectLinker::addInputFileToTar(InputFile *Ipt, MappingFile::Kind K) {
 bool ObjectLinker::readAndProcessInput(Input *Input, bool IsPostLto) {
   if (Input->isInternal())
     return true;
-  if (!Input->getSize())
-    ThisConfig.raise(Diag::input_file_has_zero_size) << Input->decoratedPath();
+  if (!Input->getSize()) {
+    if (Input->getMemArea())
```
Comment (quic-seaswara) [inline lib/Object/ObjectLinker.cpp]: Can you add a test with archive files / thin archive files / linker scripts where input file sizes can be empty or 0.

**Review by quic-seaswara — CHANGES_REQUESTED** [review]

**Comment by deepakshirkem** [comment]
Hi @parth / @quic-areg  / @quic-seaswara , Updated the suggested test. Please review when get a chance and also  the run pipeline.

File: test/Common/standalone/EmptyInputFile/EmptyInputFile.test [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+# 1. Direct empty object file
+# 2. Archive containing empty member
+# 3. Thin archive containing empty member
+# 4. Linker script referencing empty file
+#END_COMMENT
+#START_TEST
+RUN: %clang %clangopts -o %t1.o %p/Inputs/1.c -c
+RUN: touch %t1.empty.o
+RUN: llvm-ar rcs %t1.a %t1.o %t1.empty.o
+RUN: llvm-ar rcsT %t1.thin.a %t1.o %t1.empty.o
```
Comment (quic-seaswara) [inline test/Common/standalone/EmptyInputFile/EmptyInputFile.test]: Use the lit variables, do we know if this will work on windows ?

File: test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+#BEGIN_COMMENT
+# Verify that ELD does not crash when a linker script references a
+# file that cannot be read (e.g. not found in sysroot).
+# ELD should error out gracefully instead of segfaulting.
+#END_COMMENT
+#START_TEST
+RUN: %clang %clangopts -o %t1.o %p/Inputs/main.c -c
+RUN: %not %link %linkopts %t1.o -T %p/Inputs/script.t 2>&1 | %filecheck %s
+#END_TEST
+CHECK: cannot read file
```
Comment (quic-seaswara) [inline test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: which file ?

File: lib/Object/ObjectLinker.cpp (hunk 3 above)
Comment (quic-seaswara) [reply lib/Object/ObjectLinker.cpp]: Please add a comment on when the size is 0 and has an MemArea. 

I am confused as well.

**Comment by deepakshirkem** [comment]
Hi @quic-seaswara,  addressed your comments.

**Review by quic-seaswara — APPROVED** [review]

**Comment by deepakshirkem** [comment]
Hi @quic-seaswara, Can you re-run the pipeline one more time.

**Review by quic-areg — APPROVED** [review]

**Comment by deepakshirkem** [comment]
Hi @quic-seaswara, I am not sure why the CI failed for that test case, as it is passing locally for me. I think it might be because my branch was out of date. I have now rebased it with the latest top/main branch.  Could you please run the CI one more time? Sorry

File: lib/Object/ObjectLinker.cpp [hunk 6]
```diff
@@ -3618,8 +3618,19 @@ void ObjectLinker::addInputFileToTar(InputFile *Ipt, MappingFile::Kind K) {
...
-  if (!Input->getSize())
-    ThisConfig.raise(Diag::input_file_has_zero_size) << Input->decoratedPath();
+  if (!Input->getSize()) {
+    // Two cases where size is 0:
+    // 1. File exists but is empty (MemArea != null, size = 0):
+    //    Emit a note and continue processing.
+    // 2. File does not exist (MemArea = null):
+    //    fatal_cannot_read_input already raised. Return false to
+    //    prevent crash in downstream processing.
+    if (Input->getMemArea())
```
Comment (parth-07) [inline lib/Object/ObjectLinker.cpp]: We should not reach readAndProcessInput for an invalid `Input *`. We resolve the input path and create the MemoryArea for the input in `Input::resolvePath()`. This function should report the error and return false if the file does not exist.

**Review by parth-07 — CHANGES_REQUESTED** [review]

File: lib/Object/ObjectLinker.cpp (hunk 6 above)
Comment (deepakshirkem) [reply lib/Object/ObjectLinker.cpp]: You were right. I found that the root cause is not in `resolvePath()`, but in `resolvePathMappingFile()`. The issue is that `createMemoryArea()` returns `nullptr` for the missing file, but `setMemArea(nullptr)` is still called and the function returns `true`, incorrectly signaling success to the caller. Added GDB logs to the comment as proof.

Really appreciate your review and pointing me in the right direction. Please take a look when you get a chance, and let me know if you have any additional test cases in mind.

**Comment by deepakshirkem** [comment]
Hi @quic-areg / @quic-seaswara / @parth-07 , Here is the GDB log confirming the root cause. Observation   `Input->getMemArea()` = `0x0` — null MemArea enters `readAndProcessInput()` Execution continues past the size check with no early return Crash occurs inside `ScriptLexer` when it dereferences the null `MemoryArea`

```
Breakpoint 1 (ObjectLinker::readAndProcessInput) pending.
Starting program: /home/deepak/Desktop/Deepak/eld/obj/bin/ld.eld --sysroot=A -T LinkerScript/script.t.11452439568106356749 Object/main.o.8683850154013456601 --mapping-file=mapping.ini

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, eld::ObjectLinker::readAndProcessInput (this=0x555555809c60, Input=0x55555582a330, IsPostLto=false) at /home/deepak/Desktop/Deepak/eld/lib/Object/ObjectLinker.cpp:3619
3619	  if (Input->isInternal())
Continuing.

Breakpoint 1, eld::ObjectLinker::readAndProcessInput (this=0x555555809c60, Input=0x55555582a410, IsPostLto=false) at /home/deepak/Desktop/Deepak/eld/lib/Object/ObjectLinker.cpp:3619
3619	  if (Input->isInternal())
$1 = (eld::MemoryArea *) 0x0
$2 = 0
3621	  if (!Input->getSize())
3622	    ThisConfig.raise(Diag::input_file_has_zero_size) << Input->decoratedPath();
3623	  LayoutInfo *layoutInfo = ThisModule->getLayoutInfo();
3624	  std::string Path = Input->getResolvedPath().native();
3626	  InputFile *CurInput = Input->getInputFile();
3627	  if (!CurInput) {
3628	    InputFile *I = InputFile::create(Input, ThisConfig.getDiagEngine());
3629	    if (!I) {
3635	    CurInput = I;
3636	    Input->setInputFile(I);
3638	  if (CurInput->shouldSkipFile()) {
3644	  if (Input->getAttribute().isPatchBase() &&
3652	  if (CurInput->isBinaryFile()) {
3867	  else if (CurInput->getKind() == InputFile::GNULinkerScriptKind) {
3868	    eld::RegisterTimer T("Read Linker Script", "Read all Input files",
3869	                         ThisConfig.options().printTimingStats());
3868	    eld::RegisterTimer T("Read Linker Script", "Read all Input files",
3870	    if (layoutInfo)
3872	    addInputFileToTar(CurInput, eld::MappingFile::LinkerScript);
3873	    CurInput->setToSkip();
3874	    if (!readLinkerScript(CurInput)) {

Program received signal SIGSEGV, Segmentation fault.
0x00007fffedde2f45 in std::__uniq_ptr_impl<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >::_M_ptr (this=0x0) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:199
199	      pointer    _M_ptr() const noexcept { return std::get<0>(_M_t); }
SignalHandler (Sig=0, Info=0x7fffffffa5d8, Context=0x7fffedde2f45 <std::__uniq_ptr_impl<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >::_M_ptr() const+21>) at /home/deepak/Desktop/Deepak/llvm-project/llvm/lib/Support/Unix/Signals.inc:399
399	static void SignalHandler(int Sig, siginfo_t *Info, void *Context) {
404	  sys::unregisterHandlers();
#0  SignalHandler (Sig=11, Info=0x5555557e8bb0, Context=0x5555557e8a80) at /home/deepak/Desktop/Deepak/llvm-project/llvm/lib/Support/Unix/Signals.inc:404
#1  <signal handler called>
#2  0x00007fffedde2f45 in std::__uniq_ptr_impl<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >::_M_ptr (this=0x0) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:199
#3  0x00007fffedde2f25 in std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >::get (this=0x0) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:470
#4  0x00007fffef7706ed in std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer> >::operator* (this=0x0) at /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/unique_ptr.h:453
#5  0x00007ffff0b5dd0d in eld::MemoryArea::getMemoryBufferRef (this=0x0) at /home/deepak/Desktop/Deepak/eld/include/eld/Support/MemoryArea.h:63
#6  0x00007ffff0b5d770 in eld::Input::getMemoryBufferRef (this=0x55555582a410) at /home/deepak/Desktop/Deepak/eld/include/eld/Input/Input.h:115
#7  0x00007ffff0d2c4f8 in eld::v2::ScriptLexer::ScriptLexer (this=0x7fffffffa7a8, Config=..., ScriptFile=...) at /home/deepak/Desktop/Deepak/eld/lib/ScriptParser/ScriptLexer.cpp:82
#8  0x00007ffff0d09439 in eld::v2::ScriptParser::ScriptParser (this=0x7fffffffa7a8, Config=..., File=...) at /home/deepak/Desktop/Deepak/eld/lib/ScriptParser/ScriptParser.cpp:53
#9  0x00007ffff0c186ab in eld::ScriptReader::readLinkerScriptUsingNewParser (this=0x555555813d00, PConfig=..., PScriptFile=...) at /home/deepak/Desktop/Deepak/eld/lib/Script/ScriptReader.cpp:47
#10 0x00007ffff0c18675 in eld::ScriptReader::readScript (this=0x555555813d00, PConfig=..., PScriptFile=...) at /home/deepak/Desktop/Deepak/eld/lib/Script/ScriptReader.cpp:42
#11 0x00007ffff0af45d4 in eld::ObjectLinker::readLinkerScript (this=0x555555809c60, Input=0x555555835080) at /home/deepak/Desktop/Deepak/eld/lib/Object/ObjectLinker.cpp:228
#12 0x00007ffff0af66d2 in eld::ObjectLinker::readAndProcessInput (this=0x555555809c60, Input=0x55555582a410, IsPostLto=false) at /home/deepak/Desktop/Deepak/eld/lib/Object/ObjectLinker.cpp:3874
#13 0x00007ffff0b5c69c in eld::GroupReader::readGroup (this=0x555555811ce0, CurNode=0x55555582b3d0, PBuilder=..., PConfig=..., IsPostLtoPhase=false) at /home/deepak/Desktop/Deepak/eld/lib/Object/GroupReader.cpp:55
#14 0x00007ffff0af488f in eld::ObjectLinker::readInputs (this=0x555555809c60, InputVector=std::vector of length 3, capacity 4 = {...})
```

The fix is in `Input::resolvePathMappingFile()`  adding a null check after `createMemoryArea()` to return `false` early instead of propagating a null `MemArea` downstream. Please let me know if you have any additional test cases in mind! Thank You ::((

**Comment by deepakshirkem** [comment]
Hi @quic-seaswara / @quic-areg / @parth-07 ,  Does this need changes or can it be merged?

**Review by quic-areg — APPROVED** [review]

**Review by Steven6798 — APPROVED** [review]
Looks good to me

**Review by parth-07 — CHANGES_REQUESTED** [review]
Please change the commit message and PR description to: `Fix reproducer crash when input file has no contents`.

**Comment by deepakshirkem** [comment]
> Please change the commit message and PR description to: `Fix reproducer crash when input file has no contents`.

Done.

**Comment by parth-07** [comment]
@deepakshirkem I do not see any update to the commit title.

**Comment by deepakshirkem** [comment]
> @deepakshirkem I do not see any update to the commit title.

Done! Changes have been pushed. Thank You ::((

File: test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test [hunk 7]
```diff
@@ -0,0 +1,11 @@
+#---SysrootScriptCrash.test----------- Executable -----------------#
```
Comment (parth-07) [inline test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: @deepakshirkem This test passes even without the changes of this patch, right? Why did we add this test with this patch?

Comment (deepakshirkem) [reply test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: Hi @parth-07, Updated.

Comment (parth-07) [reply test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: Can you please reply to the question asked as well? Why did you add this test with the patch?

Comment (deepakshirkem) [reply test/Common/standalone/SysrootScriptCrash/SysrootScriptCrash.test]: Hi @parth-07 , apologies for not answering  your question directly earlier. You are correct `SysrootScriptCrash` passes without this patch. I mistakenly added it thinking it tested the fix, but after verifying, it goes through `resolvePath() ` not `resolvePathMappingFile()`, so it does not exercise our fix at all.

I have now removed it. The correct test is `MissingMappingFileInput` which uses  --mapping-file and fails without the patch,  directly testing the fix in  `resolvePathMappingFile()`.

Thank You ::((

**Comment by deepakshirkem** [comment]
Hi @quic-seaswara / @parth-07 , This need changes or can it be merged?

File: test/Common/standalone/EmptyInputFile/EmptyInputFile.test
Comment (parth-07) [inline test/Common/standalone/EmptyInputFile/EmptyInputFile.test:1]: Do we need the fix added by this PR for this test to pass? If not, then this test should not be part of this PR.

File: test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+# ELD should report a fatal error instead of crashing with a segfault.
+#END_COMMENT
+#START_TEST
+RUN: %clang %clangopts -c %p/Inputs/main.c -o %t1.o
+RUN: %not %link %linkopts %t1.o \
+RUN:   -T %p/Inputs/missing.t \
+RUN:   --mapping-file=%p/Inputs/mapping_missing.ini \
+RUN:   2>&1 | %filecheck %s
+#END_TEST
+CHECK: cannot read file
```
Comment (parth-07) [inline test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test]: Can you add the filename in the CHECK diagnostic as well? I am concerned that we are printing here the hashed filename, instead of the original filename that the user is aware of.

File: test/Common/standalone/EmptyInputFile/EmptyInputFile.test
Comment (deepakshirkem) [reply test/Common/standalone/EmptyInputFile/EmptyInputFile.test:1]: https://github.com/qualcomm/eld/pull/1166#discussion_r3230467463   Reference

Comment (parth-07) [reply test/Common/standalone/EmptyInputFile/EmptyInputFile.test:1]: This was because the (below) previous fix was misleading. 

```diff
-  if (!Input->getSize())
-    ThisConfig.raise(Diag::input_file_has_zero_size) << Input->decoratedPath();
+  if (!Input->getSize()) {
+    if (Input->getMemArea())
```

If I understand correctly, the current fix is not affecting how zero sized input files are handled.

Comment (deepakshirkem) [reply test/Common/standalone/EmptyInputFile/EmptyInputFile.test:1]: Yes, you are correct. Sorry about that. Will remove this test, address the review comments, and update the patch.

File: test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test (hunk 8 above)
Comment (deepakshirkem) [reply test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test]: Updated the check

**Comment by deepakshirkem** [comment]
Hi @parth-07, Please run the pipeline one more time when you get a chance?

**Comment by parth-07** [comment]
> Hi @parth-07, Please run the pipeline one more time when you get a chance?

Done.

**Review by quic-seaswara — APPROVED** [review]

File: test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test [hunk 9]
```diff
@@ -0,0 +1,14 @@
+#---MissingMappingFileInput.test----------- Executable -----------------#
```
Comment (parth-07) [inline test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test:1]: @deepakshirkem Can you please check if this test would pass without the changes of this PR? It seems that this PR is not testing the bug that it is actually fixing. 

The expected error of this test is `cannot find missing.t`. The mapping file is ineffective here because the mapping key is `missing.t=DOESNOTEXIST` and there is no input named `missing.t` in the link command. Please note that `%p/Inputs/missing.t` is a different than `missing.t` for the mapping file.

Additionally, the PR / commit title is different from the actual issue being resolved.

Comment (deepakshirkem) [reply test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test:1]: Hi @parth-07
 >@deepakshirkem Can you please check if this test would pass without the changes of this PR? It seems that this PR is not testing the bug that it is actually fixing.

I verified that the test fails without the changes from this PR.

>The expected error of this test is cannot find missing.t. The mapping file is ineffective here because the mapping key is missing.t=DOESNOTEXIST and there is no input named missing.t in the link command. Please note that %p/Inputs/missing.t is a different than missing.t for the mapping file.

my intention was to test `resolvePathMappingFile()` by mapping missing.t to a non-existent file.  I did not realize that %p/Inputs/missing.t expands to an absolute path, which does not match the mapping key missing.t.

>A dditionally, the PR / commit title is different from the actual issue being resolved.

would you prefer that I send a follow-up PR to fix the test and update the title/description, or should we revert this PR and resubmit it with the correct test and description?

Comment (parth-07) [reply test/Common/standalone/MissingMappingFileInput/MissingMappingFileInput.test:1]: Can you please revert the PR and submit it with correct test and description?
